File: coins/src/main/java/com/bubelov/coins/dagger/AppComponent.java
package com.bubelov.coins.dagger;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;

import com.bubelov.coins.MerchantsCache;
import com.bubelov.coins.api.CoinsApi;
import com.bubelov.coins.provider.CoinsProvider;
import com.bubelov.coins.util.MapMarkersCache;
import com.google.gson.Gson;

import javax.inject.Singleton;

import dagger.Component;

/**
 * Author: Igor Bubelov
 * Date: 26/03/16 17:57
 */

@Singleton
@Component(modules = {AppContextModule.class, DatabaseModule.class, CacheModule.class, ApiModule.class, ConverterModule.class})
public interface AppComponent {
    Context getContext();

    SQLiteDatabase database();

    MerchantsCache getMerchantsCache();

    MapMarkersCache getMarkersCache();

    CoinsApi provideApi();

    Gson provideGson();

    void inject(CoinsProvider provider);
    void inject(MerchantsCache cache);
}

File: app/src/main/java/com/bubelov/coins/dagger/AppComponent.kt
package com.bubelov.coins.dagger

import android.content.Context
import com.bubelov.coins.App

import com.bubelov.coins.ui.viewmodel.ExchangeRatesViewModel
import com.bubelov.coins.ui.viewmodel.MainViewModel
import com.bubelov.coins.ui.viewmodel.NotificationAreaViewModel
import com.bubelov.coins.ui.viewmodel.PlacesSearchViewModel

import javax.inject.Singleton

import dagger.Component
import dagger.BindsInstance
import dagger.android.AndroidInjectionModule

/**
 * @author Igor Bubelov
 */

@Singleton
@Component(modules = arrayOf(AppModule::class, AndroidInjectionModule::class, ActivityBuilder::class, FragmentBuilder::class, ServiceBuilder::class))
interface AppComponent {
    fun inject(app: App)

    fun inject(target: MainViewModel)
    fun inject(target: ExchangeRatesViewModel)
    fun inject(target: NotificationAreaViewModel)
    fun inject(target: PlacesSearchViewModel)

    @Component.Builder
    interface Builder {
        @BindsInstance
        fun context(context: Context): Builder
        fun build(): AppComponent
    }
}